File: bigseqkit/seq.go
package bigseqkit

import "ignis/driver/api"

type SeqKitSeqOptions struct {
	inner SeqOptions
}

type SeqOptions struct {
	Config            KitConfig
	Reverse           *bool
	Complement        *bool
	Name              *bool
	Seq               *bool
	Qual              *bool
	OnlyId            *bool
	RemoveGaps        *bool
	GapLetters        *string
	LowerCase         *bool
	UpperCase         *bool
	Dna2rna           *bool
	Rna2dna           *bool
	ValidateSeq       *bool
	ValidateSeqLength *int
	MaxLen            *int
	MinLen            *int
	QualAsciiBase     *int
	MinQual           *float64
	MaxQual           *float64
}

func (this *SeqOptions) setDefaults() *SeqOptions {
	this.Config.setDefaults()
	setDefault(&this.Reverse, false)
	setDefault(&this.Complement, false)
	setDefault(&this.Name, false)
	setDefault(&this.Seq, false)
	setDefault(&this.Qual, false)
	setDefault(&this.OnlyId, false)
	setDefault(&this.RemoveGaps, false)
	setDefault(&this.GapLetters, "- 	.")
	setDefault(&this.LowerCase, false)
	setDefault(&this.UpperCase, false)
	setDefault(&this.Dna2rna, false)
	setDefault(&this.Rna2dna, false)
	setDefault(&this.ValidateSeq, false)
	setDefault(&this.ValidateSeqLength, 10000)
	setDefault(&this.MaxLen, -1)
	setDefault(&this.MinLen, -1)
	setDefault(&this.QualAsciiBase, 33)
	setDefault(&this.MinQual, -1)
	setDefault(&this.MaxQual, -1)

	return this
}

func (this *SeqKitSeqOptions) Config(v *SeqKitConfig) *SeqKitSeqOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitSeqOptions) Reverse(v bool) *SeqKitSeqOptions {
	this.inner.Reverse = &v
	return this
}

func (this *SeqKitSeqOptions) Complement(v bool) *SeqKitSeqOptions {
	this.inner.Complement = &v
	return this
}

func (this *SeqKitSeqOptions) Name(v bool) *SeqKitSeqOptions {
	this.inner.Name = &v
	return this
}

func (this *SeqKitSeqOptions) Seq(v bool) *SeqKitSeqOptions {
	this.inner.Seq = &v
	return this
}

func (this *SeqKitSeqOptions) Qual(v bool) *SeqKitSeqOptions {
	this.inner.Qual = &v
	return this
}

func (this *SeqKitSeqOptions) OnlyId(v bool) *SeqKitSeqOptions {
	this.inner.OnlyId = &v
	return this
}

func (this *SeqKitSeqOptions) RemoveGaps(v bool) *SeqKitSeqOptions {
	this.inner.RemoveGaps = &v
	return this
}

func (this *SeqKitSeqOptions) GapLetters(v string) *SeqKitSeqOptions {
	this.inner.GapLetters = &v
	return this
}

func (this *SeqKitSeqOptions) LowerCase(v bool) *SeqKitSeqOptions {
	this.inner.LowerCase = &v
	return this
}

func (this *SeqKitSeqOptions) UpperCase(v bool) *SeqKitSeqOptions {
	this.inner.UpperCase = &v
	return this
}

func (this *SeqKitSeqOptions) Dna2rna(v bool) *SeqKitSeqOptions {
	this.inner.Dna2rna = &v
	return this
}

func (this *SeqKitSeqOptions) Rna2dna(v bool) *SeqKitSeqOptions {
	this.inner.Rna2dna = &v
	return this
}

func (this *SeqKitSeqOptions) ValidateSeq(v bool) *SeqKitSeqOptions {
	this.inner.ValidateSeq = &v
	return this
}

func (this *SeqKitSeqOptions) ValidateSeqLength(v int) *SeqKitSeqOptions {
	this.inner.ValidateSeqLength = &v
	return this
}

func (this *SeqKitSeqOptions) MaxLen(v int) *SeqKitSeqOptions {
	this.inner.MaxLen = &v
	return this
}

func (this *SeqKitSeqOptions) MinLen(v int) *SeqKitSeqOptions {
	this.inner.MinLen = &v
	return this
}

func (this *SeqKitSeqOptions) QualAsciiBase(v int) *SeqKitSeqOptions {
	this.inner.QualAsciiBase = &v
	return this
}

func (this *SeqKitSeqOptions) MinQual(v float64) *SeqKitSeqOptions {
	this.inner.MinQual = &v
	return this
}

func (this *SeqKitSeqOptions) MaxQual(v float64) *SeqKitSeqOptions {
	this.inner.MaxQual = &v
	return this
}

func Seq(input *api.IDataFrame[string], o *SeqKitSeqOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitSeqOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	libprepare, err := api.AddParam(libSource("SeqTransform"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	return api.MapPartitions[string, string](input, libprepare)
}

File: bigseqkit-py/bigseqkit/seq.py
from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitSeqOptions:

    def __init__(self):
        self.__inner = SeqOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def reverse(self, v: bool):
        self.__inner.Reverse = v

    def complement(self, v: bool):
        self.__inner.Complement = v

    def name(self, v: bool):
        self.__inner.Name = v

    def seq(self, v: bool):
        self.__inner.Seq = v

    def qual(self, v: bool):
        self.__inner.Qual = v

    def onlyId(self, v: bool):
        self.__inner.OnlyId = v

    def removeGaps(self, v: bool):
        self.__inner.RemoveGaps = v

    def gapLetters(self, v: str):
        self.__inner.GapLetters = v

    def lowerCase(self, v: bool):
        self.__inner.LowerCase = v

    def upperCase(self, v: bool):
        self.__inner.UpperCase = v

    def dna2rna(self, v: bool):
        self.__inner.Dna2rna = v

    def rna2dna(self, v: bool):
        self.__inner.Rna2dna = v

    def validateSeq(self, v: bool):
        self.__inner.ValidateSeq = v

    def validateSeqLength(self, v: int):
        self.__inner.ValidateSeqLength = v

    def maxLen(self, v: int):
        self.__inner.MaxLen = v

    def minLen(self, v: int):
        self.__inner.MinLen = v

    def qualAsciiBase(self, v: int):
        self.__inner.QualAsciiBase = v

    def minQual(self, v: float):
        self.__inner.MinQual = v

    def MaxQual(self, v: float):
        self.__inner.MaxQual = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()
        libprepare = _libSource("SeqTransform").addParam("opts", _optionsToString(opts))
        return input.mapPartitions(libprepare)


class SeqOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.Reverse = None  # bool
        self.Complement = None  # bool
        self.Name = None  # bool
        self.Seq = None  # bool
        self.Qual = None  # bool
        self.OnlyId = None  # bool
        self.RemoveGaps = None  # bool
        self.GapLetters = None  # string
        self.LowerCase = None  # bool
        self.UpperCase = None  # bool
        self.Dna2rna = None  # bool
        self.Rna2dna = None  # bool
        self.ValidateSeq = None  # bool
        self.ValidateSeqLength = None  # int
        self.MaxLen = None  # int
        self.MinLen = None  # int
        self.QualAsciiBase = None  # int
        self.MinQual = None  # float
        self.MaxQual = None  # float

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "Reverse", False)
        _setDefault(self, "Complement", False)
        _setDefault(self, "Name", False)
        _setDefault(self, "Seq", False)
        _setDefault(self, "Qual", False)
        _setDefault(self, "OnlyId", False)
        _setDefault(self, "RemoveGaps", False)
        _setDefault(self, "GapLetters", "- 	.")
        _setDefault(self, "LowerCase", False)
        _setDefault(self, "UpperCase", False)
        _setDefault(self, "Dna2rna", False)
        _setDefault(self, "Rna2dna", False)
        _setDefault(self, "ValidateSeq", False)
        _setDefault(self, "ValidateSeqLength", 10000)
        _setDefault(self, "MaxLen", -1)
        _setDefault(self, "MinLen", -1)
        _setDefault(self, "QualAsciiBase", 33)
        _setDefault(self, "MinQual", -1)
        _setDefault(self, "MaxQual", -1)

        return self


def Seq(input: IDataFrame, o: SeqKitSeqOptions = None, **kwargs):
    if o is None:
        o = SeqKitSeqOptions()
    return o._run(input, **kwargs)
